File: packages/shared/src/scenario.ts
export enum TestType {
  BEND = "bend",
  TWIST = "twist",
  TENSION = "tension",
  COMPRESSION = "compression",
}

export interface Constraint {
  region_id: string;
  vertex_indices: number[];
  position: [number, number, number];
}

export interface ForceLoad {
  region_id: string;
  vertex_indices: number[];
  force_type: TestType;
  magnitude: number;
  direction: [number, number, number];
}

export interface TestScenario {
  id: string;
  name: string;
  project_id: string;
  material_id: string;
  test_type: TestType;
  constraints: Constraint[];
  loads: ForceLoad[];
  created_at: string;
  notes: string;
}

File: apps/api/app/models/scenario.py
from __future__ import annotations

from enum import Enum
from typing import List

from pydantic import BaseModel, Field


class TestType(str, Enum):
    BEND = "bend"
    TWIST = "twist"
    TENSION = "tension"
    COMPRESSION = "compression"


class Constraint(BaseModel):
    region_id: str = Field(description="ID of the pinned face group")
    vertex_indices: List[int] = Field(default_factory=list, description="Anchored vertex indices")
    position: List[float] = Field(description="Centroid [x, y, z]")


class ForceLoad(BaseModel):
    region_id: str = Field(description="ID of the loaded face group")
    vertex_indices: List[int] = Field(default_factory=list, description="Loaded vertex indices")
    force_type: TestType
    magnitude: float = Field(ge=0, description="Force magnitude in N")
    direction: List[float] = Field(description="Normalized direction vector [x, y, z]")


class TestScenario(BaseModel):
    id: str = Field(description="Unique scenario identifier")
    name: str = Field(description="Human-readable name")
    project_id: str = Field(default="default")
    material_id: str = Field(description="Selected material profile ID")
    test_type: TestType
    constraints: List[Constraint] = Field(default_factory=list)
    loads: List[ForceLoad] = Field(default_factory=list)
    created_at: str = Field(description="ISO 8601 timestamp")
    notes: str = Field(default="")
